felix/fv/rpf_test.go

```diff
@@ -0,0 +1,178 @@
+			By("turning off RPF per device", func() {
+				felixes[0].Exec("sysctl", "-w", "net.ipv4.conf.all.rp_filter=0")
+				felixes[0].Exec("sysctl", "-w", "net.ipv4.conf.default.rp_filter=0")
+				felixes[0].Exec("sysctl", "-w", "net.ipv4.conf.eth0.rp_filter=0")
+			})
+
+			var external *workload.Workload
+
+			By("setting up node's fake external ifaces", func() {
+				// We name the ifaces ethXY since such ifaces are
+				// treated by felix as external to the node
+				//
+				// Using a test-workload creates the namespaces and the
+				// interfaces to emulate the host NICs
+
+				external = &workload.Workload{
+					Name:          "eth20",
+					C:             felixes[0].Container,
+					IP:            "192.168.20.1",
+					Ports:         "57005", // 0xdead
+					Protocol:      "udp",
+					InterfaceName: "eth20",
+				}
+				err := external.Start()
+				Expect(err).NotTo(HaveOccurred())
+
+				// assign address to eth20 and add route to the .20 network
+				felixes[0].Exec("ip", "route", "add", "192.168.20.0/24", "dev", "eth20")
+				felixes[0].Exec("ip", "addr", "add", "10.0.0.20/32", "dev", "eth20")
+				_, err = external.RunCmd("ip", "route", "add", "10.0.0.20/32", "dev", "eth0")
+				Expect(err).NotTo(HaveOccurred())
+				// Add a route to the test workload to the fake external
+				// client emulated by the test-workload so that RPF can find a
+				// valid route.
+				_, err = external.RunCmd("ip", "route", "add", w.IP+"/32", "via", "10.0.0.20")
+				Expect(err).NotTo(HaveOccurred())
+
+				// Make sure that networking with the .20 network works
+				cc.ResetExpectations()
+				cc.Expect(Some, external, w)
+				cc.CheckConnectivity()
+			})
+
+			By("testing with bad source", func() {
+				fakeWorkloadIP := "10.65.15.15"
+
+				tcpdumpHEP := felixes[0].AttachTCPDump("eth20")
+				tcpdumpHEP.SetLogEnabled(true)
+				matcherHEP := fmt.Sprintf("IP %s\\.30446 > %s\\.30446: UDP", fakeWorkloadIP, w.IP)
+				tcpdumpHEP.AddMatcher("UDP-30446", regexp.MustCompile(matcherHEP))
+				tcpdumpHEP.Start()
+				defer tcpdumpHEP.Stop()
+
+				tcpdumpWl := w.AttachTCPDump()
+				tcpdumpWl.SetLogEnabled(true)
+				matcherWl := fmt.Sprintf("IP %s\\.30446 > %s\\.30446: UDP", fakeWorkloadIP, w.IP)
+				tcpdumpWl.AddMatcher("UDP-30446", regexp.MustCompile(matcherWl))
+				tcpdumpWl.Start()
+				defer tcpdumpWl.Stop()
+
+				_, err := external.RunCmd("/pktgen", fakeWorkloadIP, w.IP, "udp",
+					"--port-src", "30446", "--port-dst", "30446", "--ip-id", "666")
+				Expect(err).NotTo(HaveOccurred())
+
+				// Expect to see the packet from the .20 network at eth20 before RPF
+				Eventually(func() int { return tcpdumpHEP.MatchCount("UDP-30446") }, "1s", "100ms").
+					Should(BeNumerically("==", 1), "HEP - "+matcherHEP)
+
+				// Expect not to receive the packet from the .20 as it is dropped by RPF.
+				Consistently(func() int { return tcpdumpWl.MatchCount("UDP-30446") }, "1s", "100ms").
+					Should(BeNumerically("==", 0), "Wl - "+matcherWl)
+			})
```

Comment:
Think you need a test where you set BPFRPF to disabled and verify that you get the opposite behaviour, otherwise, we could easily break the test so that the packets don't get through for some other reason.